plugins/GSdx/GSFastList.h

```diff
@@ -0,0 +1,221 @@
+			ListRemove(index);
+			ListInsertFront(index);
+		}
+	}
+
+	__forceinline const FastListIterator<T> begin() const {
+		return FastListIterator<T>(this, FirstIndex());
+	}
+
+	__forceinline const FastListIterator<T> end() const {
+		return FastListIterator<T>(EndIndex());
+	}	
+
+private:
+	// Accessed by FastListIterator<T> using class friendship
+	__forceinline const T Data(const uint16 index) const {
+		return m_buffer[index].data;
+	}
+
+	// Accessed by FastListIterator<T> using class friendship
+	__forceinline uint16 NextIndex(const uint16 index) const {
+		return m_buffer[index].next_index;
+	}
+
+	__forceinline uint16 FirstIndex() const {
+		return m_buffer[0].next_index;
+	}
+
+	__forceinline uint16 EndIndex() const {
+		return 0;
+	}
+	
+	__forceinline bool Full() const {
+		return m_free_indexes_stack_top == m_capacity - 1;
+	}
+
+	__forceinline void ListInsertFront(const uint16 index) {
+		// Update prev / next indexes to add m_buffer[index] to the chain
+		Element<T>& head = m_buffer[0];
+		m_buffer[index].prev_index = 0;
+		m_buffer[index].next_index = head.next_index;
+		if (head.next_index != EndIndex()) {
+			m_buffer[head.next_index].prev_index = index;
+		}
+		head.next_index = index;
+	}
+
+	__forceinline void ListRemove(const uint16 index) {
+		// Update prev / next indexes to remove m_buffer[index] from the chain
+		const Element<T>& to_remove = m_buffer[index];
+		m_buffer[to_remove.prev_index].next_index = to_remove.next_index;
+		if (to_remove.next_index != EndIndex()) {
+			m_buffer[to_remove.next_index].prev_index = to_remove.prev_index;
+		}
+	}
+
+	void Grow() {
+		if (m_capacity == USHRT_MAX) {
+			throw std::runtime_error("FastList size maxed out at USHRT_MAX (65535) elements, cannot grow futhermore.");
+		}
+
+		// 32768 is the first number that when doubled overflows uint16
+		const uint16 new_capacity = m_capacity < 32768 ? m_capacity * 2 : USHRT_MAX;
+
+		Element<T>* new_buffer = (Element<T>*)_aligned_malloc(new_capacity * sizeof(Element<T>) + (new_capacity - 1) * sizeof(uint16), 64);
+		uint16* new_free_indexes_stack = (uint16*)&new_buffer[new_capacity];
+
+		memcpy(new_buffer, m_buffer, m_capacity * sizeof(Element<T>));
+		memcpy(new_free_indexes_stack, m_free_indexes_stack, (m_capacity - 1) * sizeof(uint16));
+		
+		_aligned_free(m_buffer);
+		
+		m_buffer = new_buffer;
+		m_free_indexes_stack = new_free_indexes_stack;
+
+		// Initialize the additional space in the stack
+		for (uint16 i = m_capacity - 1; i < new_capacity - 1; i++) {
+			m_free_indexes_stack[i] = i + 1;
+		}
+
+		m_capacity = new_capacity;
+	}
+};
+
+
+template <class T>
+// This iterator is const_iterator by design as T is intended to be a primitive type or pointer
+class FastListIterator
+{
+private:
+	// Should be a * instead of &, because otherwise it cannot be uninitialized (see end())
+	const FastList<T>* m_fastlist;
+	uint16 m_index;
+
+public:
+	// Used for end() iterator, which does not need m_fastlist ptr.
```

Comment:
Oh. What do you miss ? Only the value of prev_index for the head ? Is it a big change ? I would prefer to have it.